move test components to shared library to support SM integration to biologics

labkey-chrisj opened the pull request:
#### Rationale
This moves InsertRecord.java and related dialog(s) from SM, to the shared library in support of integrating SM into biologics.
Proposed rename -> SampleCRUDPanel

#### Related Pull Requests
https://github.com/LabKey/sampleManagement/pull/320
https://github.com/LabKey/biologics/pull/628

#### Changes
Update bulk edit lookup method to use FilteringReactSelect, helpful for cases where there are many entities to resolve in the lookup

labkey-chrisj committed:
move sampleCRUDPanel to shared libary

labkey-chrisj committed:
Merge remote-tracking branch 'origin/develop' into fb_migrate_create_samples

labkey-chrisj committed:
fix off-by-one issue when getting first data column when there's a selector column

labkey-chrisj committed:
Better name for InsertEntitiesPanel

labkey-chrisj committed:
add getColumnData to EditableGrid

labkey-tchad commented on src/org/labkey/test/components/glassLibrary/components/InsertEntitiesPanel.java:
```diff
@@ -0,0 +1,193 @@
+
+// This is the same component (collection of atomic elements) used in insertAssay and createSamples.
+
+public class InsertEntitiesPanel extends WebDriverComponent<InsertEntitiesPanel.ElementCache>
```
Reference the specific product component that this class controls. Also rename to match product component (EntityInsertPanel.tsx).
```suggestion
/**
 * This class automates the UI component defined in <a href="https://github.com/LabKey/labkey-ui-components/blob/master/packages/components/src/components/entities/EntityInsertPanel.tsx">components/entities/EntityInsertPanel.tsx</a>
 * This is the same component (collection of atomic elements) used in insertAssay and createSamples.
 */
public class EntityInsertPanel extends WebDriverComponent<EntityInsertPanel.ElementCache>
```

labkey-danield commented on src/org/labkey/test/components/react/dialogs/BulkCreateSamplesDialog.java:
```diff
@@ -0,0 +1,159 @@
+    public BulkCreateSamplesDialog setSelectionField(String fieldCaption, List<String> selectValues)
+    {
+        FilteringReactSelect reactSelect = FilteringReactSelect.finder(getDriver()).followingLabelWithSpan(fieldCaption).find();
+        getWrapper().waitFor(()-> reactSelect.isInteractive(), 5000);
```
Are you sure this will work?
My experience was that the arrow on the reactSelect was showing up enabled when data was still loading. That is why the waitForReactSelectReady was added.

labkey-chrisj replied:
I'm willing to be wrong on this.  But, if waitForReactSelectReady works, I'd be happy to implement a real waitForLoaded method on BaseReactSelect.  

After a bit of investigation, it's clear that `.isInteractive()` was the wrong condition to check. I'm implementing a public 'waitForLoaded' method on BaseReactSelect that uses both the absence of a loading spinner and of the "Loading..." value placeholder to determine whether the control is still loading options.

labkey-danield replied:
Ok, that sounds like a good approach. The call to .isInteractive() is not a reliable check if the dropdown control is a look up field. This is why the [SMSampleCreateTest.testBulkInsert](https://teamcity.labkey.org/buildConfiguration/LabKey_Trunk_Premium_SampleManager_SampleManagerAPostgres/1055312?buildTab=tests&pager.currentPage=1) is failing in the branch. A call to isInteractive is returning true but there is a split second between that and the list int he drop down being populated, which is enough time for the test code to type the text which is not found.

labkey-chrisj replied:
At the moment, SMSampleCreateTest.testBulkInsert is failing for a different reason.
By switching to use FilteringReactSelect.filterSelect(), the test filters the input, and looks for a value that contains the filter clause- but values here don't contain the full filter being supplied.  Easy workaround: omit the Color: prefix, just search for the color ("orange")
I've sent you a patch with this workaround, wanted to give you the input on whether it's the right fix

labkey-martyp committed:
Case insensitive query select in schema browser

GetQueryDetailsAction now returns correct casing of names.  Many tests are not using correct casing when selecting queries in schema browser.

labkey-tchad commented on src/org/labkey/test/components/glassLibrary/components/InsertEntitiesPanel.java:
```diff
@@ -0,0 +1,193 @@
+package org.labkey.test.components.glassLibrary.components;
```
Avoid using the `glassLibrary` package. Use `org.labkey.test.components.labkey.ui`
```suggestion
package org.labkey.test.components.labkey.ui;
```

labkey-chrisj committed:
Merge remote-tracking branch 'origin/develop' into fb_migrate_create_samples

labkey-chrisj committed:
give ReactSelect the ability to wait for loaded

labkey-tchad commented on src/org/labkey/test/components/glassLibrary/components/BaseReactSelect.java:
```diff
@@ -90,6 +90,18 @@ public boolean isLoading()
+
     public String getValue()
     {
         return getWrapper().shortWait().withMessage(() -> "Select stuck loading. [" + getComponentElement().getText() + "]").until(ignored -> {
```
There are too many different methods for determining if the select has loaded. `isLoading` and `isLoaded` are redundant; then `getValue` and `isExpanded` do their own thing.

labkey-tchad commented on src/org/labkey/test/components/glassLibrary/components/BaseReactSelect.java:
```diff
@@ -90,6 +90,18 @@ public boolean isLoading()
+        return !isLoading() && !getOptions().contains("Loading...");
+    }
+
+    public T waitForLoaded(int waitMsec)
+    {
+        _wrapper.waitFor(()-> isLoaded(),
+                "Took too long for to become loaded", waitMsec);
+        return (T) this;
+    }
```
I don't see a reason for any of the loaded/loading methods to be public. Any interaction methods should handle waiting internally.
```suggestion
    protected void waitForLoaded(int waitMsec)
    {
        _wrapper.waitFor(()-> isLoaded(),
                "Took too long for select to load: " + getComponentElement().toString(), waitMsec);
    }
```

labkey-chrisj replied:
Okay.  I'll make waitForLoaded() protected and call it from interaction/entry points
If the logic is different in derived classes we can override there

labkey-tchad commented on src/org/labkey/test/components/react/dialogs/BulkCreateSamplesDialog.java:
```diff
@@ -0,0 +1,160 @@
+package org.labkey.test.components.react.dialogs;
```
This is a LabKey developed component
```suggestion
org.labkey.test.components.labkey.ui.samples;
```

labkey-chrisj replied:
done

labkey-tchad commented on src/org/labkey/test/components/react/dialogs/BulkCreateSamplesDialog.java:
```diff
@@ -0,0 +1,160 @@
+        return getWrapper().getFormElement(Locator.tagWithId("input", forField));
+    }
+
+    private boolean waitForReactSelectReady(FilteringReactSelect select)
+    {
+        getWrapper().waitFor(()->!select.getSelections().contains("Loading..."), "The reactSelect field did not populate in time.", 1_000);
+        return true;
+    }
```
This component shouldn't need to determine whether the select is ready and it certainly shouldn't need to know how to do so.

labkey-danield replied:
I agree. This was a hack/work-around for an issue with the ReactSelect. I'll start an email thread to discuss best solutions.

labkey-tchad commented on src/org/labkey/test/components/react/dialogs/BulkCreateSamplesDialog.java:
```diff
@@ -0,0 +1,160 @@
+        }
+    }
+
+    @Override
+    protected ElementCache newElementCache()
+    {
+        return (ElementCache) super.elementCache();
+    }
+
+    @Override
+    protected ElementCache elementCache()
+    {
+        return new ElementCache();
+    }
```
You have these backward.
```suggestion
    @Override
    protected ElementCache newElementCache()
    {
        return new ElementCache();
    }

    @Override
    protected ElementCache elementCache()
    {
        return (ElementCache) super.elementCache();
    }
```

labkey-chrisj replied:
done

labkey-tchad commented on src/org/labkey/test/components/react/dialogs/BulkCreateSamplesDialog.java:
```diff
@@ -0,0 +1,160 @@
+
+public class BulkCreateSamplesDialog extends ModalDialog
+{
+
+    public BulkCreateSamplesDialog(WebDriver driver)
+    {
+        this(new ModalDialogFinder(driver).withTitle("Bulk Creation of Samples"));
+    }
+
+    private BulkCreateSamplesDialog(ModalDialogFinder finder)
+    {
+        super(finder);
+    }
```
Take an `InsertEntitiesPanel` in the constructor. That way this component could wait for the list to update in `clickAddRows`.
```suggestion
    private final InsertEntitiesPanel _panel;

    public BulkCreateSamplesDialog(InsertEntitiesPanel panel)
    {
        super(new ModalDialogFinder(panel.getDriver()).withTitle("Bulk Creation of Samples"));
        _panel = panel;
    }
```
`InsertEntitiesPanel.getDriver()` needs to be public for this to work (I really wish I'd made `WebDriverComponent.getDriver` public in the first place 🤦)

labkey-chrisj replied:
BCSD now takes an InsertEntitesPanel in its constructor, getDriver() is now public on InsertEntitiesPanel

labkey-tchad commented on src/org/labkey/test/components/react/dialogs/BulkCreateSamplesDialog.java:
```diff
@@ -0,0 +1,160 @@
+        {
+            // Do nothing if stale.
+        }
+    }
```
Wait for dialog to disappear. Maybe wait for the insert page to update somehow as well.
```suggestion
        waitForClose();
        /* Wait for insert panel to update */
    }
```

labkey-tchad commented on src/org/labkey/test/components/react/dialogs/BulkCreateSamplesDialog.java:
```diff
@@ -0,0 +1,160 @@
+        {
+            // Do nothing if stale.
+        }
+    }
```
Wait for dialog to disappear.
```suggestion
        waitForClose();
    }
```

labkey-chrisj replied:
I've added a call to waitForClose() to the cancel and clickInsertRows methods (right after clicking the buttons, which assumes that the checks for errors are looking for errors in the EntityInsertPane (and not in the dialog)

labkey-danield committed:
Merge branch 'develop' into fb_migrate_create_samples

labkey-chrisj committed:
CR feedback
- refactor reactSelect loading checks
- rename EntityInsertPanel
- update BulkCreateSamplesDialog

labkey-danield commented on src/org/labkey/test/components/glassLibrary/components/BaseReactSelect.java:
```diff
@@ -86,19 +85,27 @@ public boolean hasValue()
-        return Locators.loadingSpinner.findElementOrNull(getComponentElement()) != null;
+        // if either are present, we're loading options
+        return Locators.loadingSpinner.existsIn(getComponentElement()) ||
+                getComponentElement().getText().equals("Loading...");
```
Sure you don't want to make this check case insensitive, and not dependent upon the ellipse?
`getComponentElement().getText().toLowerCase().contains("loading")`

labkey-chrisj replied:
A quick survey of dev code shows that yes, there are cases where lower-case is being used, so I'm down with lower-case.
The ellipsis seems to be universal, however, and I'm worried that `.toLowerCase().contains("loading")` could match actual values.  I propose to do `getComponentElement().getText().toLowerCase().equals("loading...")`.

labkey-danield commented on src/org/labkey/test/components/glassLibrary/components/BaseReactSelect.java:
```diff
@@ -86,19 +85,27 @@ public boolean hasValue()
+        waitForLoaded();
+
+        String val = getComponentElement().getText();
+        if (val.equals("Loading..."))
```
Same question as above for case insensitive and the ellipse.
However if you do want to use `equals("Loading...")` you may want to consider making "Loading..." a constant.

labkey-tchad replied:
This `if/else` shouldn't even be necessary now that we call `waitForLoaded` at the start of this method.

labkey-tchad commented on src/org/labkey/test/components/ui/samples/BulkCreateSamplesDialog.java:
```diff
@@ -0,0 +1,155 @@
+
+import java.util.List;
+
+public class BulkCreateSamplesDialog extends ModalDialog
```
Sorry I didn't notice this earlier. This should be more generic. The panel that pops this dialog is for any kind of entity. I don't think it needs any functional changes, just a rename and a move to a different package:
`org.labkey.test.components.ui.EntityBulkInsertDialog`
Should also update the title passed to the `ModalDialogFinder` in the constructor. Based on the [product code](https://github.com/LabKey/labkey-ui-components/blob/master/packages/components/src/components/entities/EntityInsertPanel.tsx#L779), `"Bulk Creation of"` should do the trick.

labkey-chrisj replied:
I think this is a good catch.  I'll do it.

labkey-chrisj committed:
cr feedback
- rename EntityInsertPanel
- pull 'loading...' out as a constant

labkey-chrisj committed:
cr feedback
- pull parent and target type selection out of page, up into EntityInsertPanel

labkey-chrisj committed:
Merge remote-tracking branch 'origin/develop' into fb_migrate_create_samples

labkey-martyp committed:
Merge branch 'develop' into fb_migrate_create_samples

labkey-martyp committed:
Test tests around public schema name

labkey-martyp committed:
Add vewQueryData method for different public name
